src/librustc_privacy/lib.rs

```diff
@@ -508,39 +508,45 @@ impl EmbargoVisitor<'tcx> {
-            if let Some(def_kind) = self.tcx.def_kind(item_def_id) {
-                let item = self.tcx.hir().expect_item(hir_id);
-                let vis = ty::Visibility::from_hir(&item.vis, hir_id, self.tcx);
-                self.update_macro_reachable_def(hir_id, def_kind, vis, defining_mod);
+    fn update_macro_reachable_mod(&mut self, reachable_mod: hir::HirId, defining_mod: DefId) {
+        let set_vis = |this: &mut Self, hir_id: hir::HirId| {
+            let item_def_id = this.tcx.hir().local_def_id(hir_id);
+            if let Some(def_kind) = this.tcx.def_kind(item_def_id) {
+                let item = this.tcx.hir().expect_item(hir_id);
+                let vis = ty::Visibility::from_hir(&item.vis, hir_id, this.tcx);
+                this.update_macro_reachable_def(hir_id, def_kind, vis, defining_mod);
             }
-        }
+        };
 
-        if let Some(exports) = self.tcx.module_exports(module_def_id) {
-            for export in exports {
-                if export.vis.is_accessible_from(defining_mod, self.tcx) {
-                    if let Res::Def(def_kind, def_id) = export.res {
-                        let vis = def_id_visibility(self.tcx, def_id).0;
-                        if let Some(hir_id) = self.tcx.hir().as_local_hir_id(def_id) {
-                            self.update_macro_reachable_def(
-                                hir_id,
-                                def_kind,
-                                vis,
-                                defining_mod,
-                            );
+        let module_def_id = self.tcx.hir().local_def_id(reachable_mod);
+        if let Some((module, _, _)) = self.tcx.hir().get_if_module(module_def_id) {
+            for item_id in &module.item_ids {
+                let hir_id = item_id.id;
+                set_vis(self, hir_id);
+            }
+            if let Some(exports) = self.tcx.module_exports(module_def_id) {
+                for export in exports {
+                    if export.vis.is_accessible_from(defining_mod, self.tcx) {
+                        if let Res::Def(def_kind, def_id) = export.res {
+                            let vis = def_id_visibility(self.tcx, def_id).0;
+                            if let Some(hir_id) = self.tcx.hir().as_local_hir_id(def_id) {
+                                self.update_macro_reachable_def(
+                                    hir_id,
+                                    def_kind,
+                                    vis,
+                                    defining_mod,
+                                );
+                            }
                         }
                     }
                 }
             }
+        } else if let Some(hir::Node::Item(hir::Item {
+            hir_id,
+            ..
+        })) = self.tcx.hir().get_if_local(module_def_id) { // #63164
+            // `macro` defined inside of an item is only visible inside of that item's scope.
+            set_vis(self, *hir_id);
```

Comment:
I'm not sure this line is doing something meaningful.
I'm also not sure how a macro in a nameless block ended up reachable at all.
Need to think a bit and re-read the code around.

Reply:
I think the issue is here:
https://github.com/rust-lang/rust/blob/9dd5c191993aab6c2f1538eb8ab69afdc4b6e67a/src/librustc_privacy/lib.rs#L890-L894

This should return early when `module_id` doesn't refer to a module.

Reply:
Yeah, this should be the right fix, we can bail out from `fn visit_macro_def` early if the macro's parent is not a module.

(Unless it's `#[rustc_macro_transparency = "opaque"] #[macro_export] macro_rules! { ... }` inside an unnamed block, but anything involving `rustc_*` attributes is entirely internal and not publicly supported.)

Reply:
@petrochenkov do you want to implement that change in a separate PR?

Reply:
It's probably better to update this PR.